bitburner-official/bitburner-src
# MISC: BitNode options

This PR adds a new feature: BitNode options. When the player enters a BN, they can set custom BitNode options:
- Active level of SF: change the effective level of SFs. For example, if the player has SF12.1000, they can set a custom level and enter a BN without NFG-1000. It also allows the player to create custom "scenarios", e.g., "I have all SFs now. How long do I need to beat BN1 with only SF1, SF4, and SF5?".
- Disable some features. These options are based on the conditions of BN challenges.

Intentional changes:
- `knowsAboutBitverse` is renamed to `knowAboutBitverse` and moved to `BitNodeUtils.ts`. It replaces all checks of `Player.sourceFiles.size > 0`.
- `canAccessBitNodeFeature` is moved to `BitNodeUtils.ts`. It replaces all checks of `Player.bitNodeN === n || Player.sourceFileLvl(n) > 0`.
- All usages of `Player.sourceFileLvl` and `Player.activeSourceFileLvl` are intentional.
- `gainIntelligenceExp`: Remove `this.skills.intelligence > 0`. This check comes from the ancient commit e3991b8795fc9bccd38240e17c99e6d20d53fd4f. It's irrelevant now. The player cannot earn intelligence exp if they are not in BN5 and do not have SF5.
- The refactor of `Sleeve.ts` and `Person.ts`: I moved functions in `[X]Methods.ts` to `[X].ts` when I started to implement the `disableSleeveExpAndAugmentation` option. That change is unnecessary now, so I will revert it if the maintainers want.

### Test case
Normal gameplay: Check SF level and active SF level after entering a new BN:
- Finish the BN:
  - Hack WD via Terminal.
  - `ns.singularity.destroyW0r1dD43m0n`.
  - Bladeburner.
- Bitflume:
  - b1t_flum3.exe.
  - `ns.singularity.b1tflum3`.

Change active SF level:
- Check SF level and active SF level after entering a new BN.
- Check effects of SF. For example, set SF3 to level 0, and check if we can create a corporation outside of BN3.

Change boolean options:
- Restrict Home PC upgrade:
  - Check UI of vendor.
  - `ns.singularity.upgradeHomeRam` and `ns.singularity.upgradeHomeCores`.
- Disable XYZ: Check if XYZ is available. For example:
  - Enter BN3, and check if we can create a corporation.
  - Set SF3 to level 3, enter BN1, and check if we can create a corporation.
- Disable 4S Market Data:
  - Check UI of Stock Market.
  - `ns.stock.purchase4SMarketData` and `ns.stock.purchase4SMarketDataTixApi`.
- Disable Sleeves' exp and augmentation:
  - Check UI of Sleeves.
  - Functions in `ns.formulas.work`.

### Screenshot

![1](https://github.com/bitburner-official/bitburner-src/assets/152669316/7166ddb3-e8f6-4153-bf07-22e49f6175b5)
![2](https://github.com/bitburner-official/bitburner-src/assets/152669316/58623f0c-74a0-4eaf-b081-cad3b06c4f93)
![3](https://github.com/bitburner-official/bitburner-src/assets/152669316/c0a3e62a-8a0f-4208-b30d-577070277c72)
